server/src/Lib.hs

```diff
@@ -0,0 +1,187 @@
+        buildTriplet appDB (Triplet bin pkg sys)
+
+-- | Check the Nix programs database for a package that provides the given binary
+-- If there is a candidate package with the same name as the binary, that package is given priority.
+-- If not, the alphabetically first package is chosen.
+resolvePackageName :: ProgramDB -> BinaryName -> Platform -> ExceptT String IO PackageName
+resolvePackageName (ProgramDB dbpath) bin sys = do
+  pkgs <- liftIO . withConnection dbpath $ \conn ->
+    query conn "SELECT package FROM programs WHERE binary = ? AND platform = ?" (bin, show sys)
+  case fmap fromOnly pkgs of
+    [] -> throwError $ "No known package provides " <> unBinaryName bin <> " for " <> show sys <> ". Consider manually specifying the package."
+    candidates -> pure $ if coerce bin `elem` candidates then coerce bin else minimum candidates
+
+-- | 1. See if we know this triplet to be unbuildable
+--   2. If buildable, build it
+--   3. Bump the count
+buildTriplet :: ApplicationDB -> Triplet -> ExceptT String IO ByteString
+buildTriplet appDB trip@(Triplet bin pkg sys) = withApplicationDB appDB $ \conn -> do
+  errorFlags :: [Only Bool] <- liftIO $ query conn "SELECT error FROM binaries WHERE binary = ? AND package = ? AND platform = ?" (bin, pkg, show sys)
+  case errorFlags of
+    [Only True] -> do
+      bump conn
+      -- TODO
+      -- Right now, upon a cached failure, we can't tell the user any more than this.
+      -- Maybe we should store some more information about where the error came from, or a part of the nix log?
+      throwError "Known failure"
+    [] -> do
+      -- TODO thread safety issue: we have an unguarded insert with a primary key, this throws an exception if in the meantime someone else makes the same request
+      -- TODO maybe just use an atomic SQL transaction is enough?
+      res <- liftIO $ nixBuild trip
+      liftIO $
+        execute
+          conn
+          "INSERT INTO binaries (binary, package, platform, hits, error) VALUES (?,?,?,?,?)"
+          (bin, pkg, show sys, 1 :: Int, isLeft res)
+      ExceptT $ pure res -- equiv to either throwError pure
+    [Only False] -> do
+      -- TODO Don't invoke nix?
+      res <- liftIO $ nixBuild trip
+      case res of
+        Left _ -> error "what to do" -- TODO
+        Right bs -> bs <$ bump conn
+    _ : _ : _ -> error "impossible"
+  where
+    -- TODO is this the right way to do thread safety? or should we take more care on the application side?
+    bump conn = liftIO $ execute conn "UPDATE binaries SET hits = hits + 1 WHERE binary = ? AND package = ? AND platform = ?" (bin, pkg, show sys)
+
+nixBuild :: Triplet -> IO (Either String ByteString)
+nixBuild (Triplet bin pkg sys) = runExceptT $ do
+  -- TODO stream build process to stderr
+  -- TODO make sure we can't get any XSS-like shenanigans
+  (exit, stdout, stderr) <- liftIO $ Proc.readProcessWithExitCode "nix" ["build", "--no-link", "--print-out-paths", packageString ""] ""
```

Comment:
```suggestion
  (exit, stdout, stderr) <- liftIO $ Proc.readProcessWithExitCode
      "nix"
      [ "build"
      , "--no-link"
      , "--print-out-paths"
      -- Only allow access to files from NIX_PATH. Derivations can't read files like `/etc/shadow`.
      , "--option", "restrict-eval", "true"
      -- Disallow IFD.  Probably not necessary, since Nixpkgs doesn't contain derivations that do IFD, but better to be on the safe side.
      , "--option", "allow-import-from-derivation", "false"
      -- Make sure the sandbox is always used.
      , "--option", "sandbox-fallback", "false"
      , packageString ""
      ]
      ""
```

Added some `nix` options just for a little extra safety.  (Although I imagine that if we run an auto-formatter, this code will get messed up a little.)

Reply:
This became 2ec46ea972546dd55207a7224edd54f11e201884